feat: Added generic client hooks for HTTP based authentication, and improved agent.json resolution

mprynce opened the pull request:
This PR adds a generic client hook for HTTP based authentication.  This allows protocols, like the [Agentic Profile](https://agenticprofile.ai), to implement Decentralized ID (DID)/JWT based authentication which then enables user, business, and gov scoped identity for A2A agents.  When an authHandler is provided and an HTTP 401 is received, the client will try a second time (see the A2AClient._fetch() function) with revised auth headers if new ones are made available by the authHandler.

This PR also improves the client agent card resolution.  Previously the card was always found at <agentpath>/.well-known/agent.json which is only correct for agents at the URL path root.  For agents with paths, such as https://example.com/agents/coder the URL for the agent card should be https://example.com/agents/coder/agent.json. This PR fixes the error.

I've moved the previous Eliza sample agent to a separate PR into a2a-samples.  The Eliza sample provides A2A server side authentication, and a modified A2A CLI with authentication.

The Eliza sample code relies on this PR to function.

swapydapy commented on src/client/client.ts:
```diff
@@ -51,9 +57,10 @@ export class A2AClient {
    * @param agentBaseUrl The base URL of the A2A agent (e.g., https://agent.example.com).
    */
-  constructor(agentBaseUrl: string) {
+  constructor(agentBaseUrl: string, options?: A2AClientOptions) {
```
I suggest, in the options, we allow users to pass on their implementation of "fetch". This would give them maximum flexibility on how to handle API auth failures. Something like:

 customFetch = (input: RequestInfo | URL, init?: RequestInit): Promise<Response>

mprynce replied:
I used the pattern from the older code of fetchImpl and passed as an option to the constructor.  Am writing unit tests next...

swapydapy commented:
Thank you for the PR. Suggested an alternative approach of allowing custom `fetch` to passed on. Please rebase with latest changes.

mprynce commented:
> Thank you for the PR. Suggested an alternative approach of allowing custom `fetch` to passed on. Please rebase with latest changes.

Good idea and happy to implement.  That was actually in the original implementation (before me) and wasn't added for the current one.  I'll get it back in :)

mprynce commented:
Added  client unit tests and I think it just needs approval from a maintainer... Let me know of any other changes that are needed.  Thanks!

swapydapy commented on src/index.ts:
```diff
@@ -3,5 +3,27 @@
  * Exports the common types.
  */
 
+export type { AgentExecutor } from "./server/agent_execution/agent_executor.js";
```
Instead of the main index.ts, could you export these server types in /server/index.ts, and the client one in /client/index.ts

We have 3 entrypoints now. One each for client & server, one for common API types

swapydapy commented on test/server/default_request_handler.spec.ts:
```diff
@@ -282,7 +284,9 @@ describe('DefaultRequestHandler as A2ARequestHandler', () => {
     it('sendMessage: should handle agent execution failure for non-blocking calls', async () => {
         const errorMessage = 'Agent failed!';
-        (mockAgentExecutor as MockAgentExecutor).execute.rejects(new Error(errorMessage));
+        (mockAgentExecutor as MockAgentExecutor).execute.callsFake(async () => {
```
is it any different from previous invocation of directly using `.execute.rejects(new Error(errorMessage));`

mprynce replied:
Previously it was failing in my local tests, but with the newest codebase the original version is working so I've reverted it

swapydapy commented on src/client/auth-handler.ts:
```diff
@@ -0,0 +1,31 @@
+    shouldRetryWithHeaders: (req:RequestInit, res:Response) => Promise<HttpHeaders | undefined>;
+
+    /* If the last call using the headers was successful, report back using this function. */
+    onSuccess: (headers:HttpHeaders) => Promise<void>
```
We can make this optional. I assume for majority cases, the implementation of `shouldRetryWithHeaders` would also save the new headers.

mprynce replied:
I've updated auth-handler.ts with a more complete description of the logic.  In short, shouldRetryWithHeaders usually provides an HTTP Authorization header that might or might not work.  This new header is not saved until the onSuccess() callback.  If the invalid header was saved it could lead to additional authorization requests with a bad header.  This logic is currently in production on Matchwise and working well.

mprynce replied:
onSuccessfulRetry is now optional

swapydapy commented:
Thanks for the changes. Added some minor comments. Once addressed, we can merge this PR.

swapydapy commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,963 @@
+        const authHeader = options?.headers?.['Authorization'] as string;
+        
+        // First call: no auth header, return 401
+        if (fetchCallCount === 2 && !authHeader) {
```
Can we simplify this fake fetch setup. I thin it can simply check for auth header, if not return 401, otherwise a response.

mprynce replied:
The first request is (intentionally) missing the authorization header, which causes the server to issues WWW-Authenticate with a challenge (in this case, which must be signed by the client).  This is the authentication flow used by the agentic profile.

swapydapy replied:
Since it is already missing an auth-header, the last line with "// Any other case without auth header should fail" would return the auth challenge

mprynce replied:
I reworked that whole section... much simpler now.  It still does the full test pass of missing auth -> send client challenge -> client signs challenge and reissues request -> server confirms signature.  I've moved all the auth code into a fetchImpl (as you suggested).  More on that later...

swapydapy commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,963 @@
+          status: 401,
+          headers: { 
+            'Content-Type': 'application/json',
+            'WWW-Authenticate': 'Agentic eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c'
```
Can we simply this WW-Authenticate header. Just keep it 'Agentic: <predefined-10-char value>'

swapydapy replied:
I would suggest to use well known auth schemes in this authenticate header. Something like:
Bearer client_id="some-id", authorization_uri="https://agent-path/oauth2/authorize", error="insufficient_claims", claims="call_api"

swapydapy commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,963 @@
+      await client.sendMessage(messageParams);
+      
+      // Reset call count for second request
+      fetchCallCount = 0;
```
As stated, don;t think we need to track this fetchCallCount

mprynce replied:
This is used to sequence the first response missing the auth, which provides the challenge, and the second response which makes sure the auth header is provided.  Otherwise a test that immediately passes the auth header would pass which is not what we are testing here.

swapydapy replied:
These tracking of fetchCounts makes the test brittle and susceptible to busg. The test are complex to reason about. Can we simplify the fake, by returning 401 when header is not present otherwise, some responses. Also, lets not use global fake, instead create one for each test.

swapydapy commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,963 @@
+      mockFetch.reset();
+      
+      // Create a new mock for the second request that expects auth header
+      mockFetch.callsFake(async (url: string, options?: RequestInit) => {
```
why is this new fake needed

mprynce replied:
(Not me, this was generated, but it seems to make sense... :)

- Original mock: Simulates the full auth flow (no token → 401 → get token → retry with token)
- New mock: Simulates a scenario where the client already has a token and should use it immediately

The new fake ensures that the test properly validates that the client is reusing the cached authentication token instead of going through the full authentication flow again.

The new fake is needed because the test is transitioning from testing the authentication flow to testing token reuse. Here's what's happening:

## The Original Mock (lines 67-180)
The original mockFetch was designed to simulate the complete authentication flow:
1. First call: Agent card fetch (always succeeds)
2. Second call: RPC request without auth header → returns 401 to trigger auth
3. Third call: RPC request with auth header → returns success
The original mock uses fetchCallCount to track which call it is and behaves differently based on that count.

## The Problem with Reusing the Original Mock
When the second test (should reuse authentication token for subsequent requests) runs:
1. The first request completes the auth flow and stores the token
2. The test calls mockFetch.reset() to clear the call count
3. But the original mock logic still depends on fetchCallCount - it expects the second API call to be call #2 (without auth) and the third to be call #3 (with auth)

## Why the New Fake is Required
The new fake on line 270 creates a different behavior that's appropriate for the second test:
1. It doesn't use fetchCallCount - it doesn't need to track call order
2. It expects the auth header immediately - since the client should now have a cached token
3. It only makes one call - no retry should be needed since the token is already available
4. It validates the exact token - checks that the correct cached token is being used

mprynce replied:
Simplified code while retaining defensive code. Tested without .reset() and .callsFake and unit tests fail.

Updated test to verify token generated from first call is used in second call.

swapydapy commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,963 @@
+
+      // Create a new mock that handles concurrent requests properly
+      mockFetch.reset();
+      mockFetch.callsFake(async (url: string, options?: RequestInit) => {
```
This new fake is not needed.

mprynce replied:
I'm working on a big refactor to make the tests much easier :)

swapydapy reviewed (changes requested):
Please go through the tests once. There seem to be lot of redundant code, especially regarding the fake Fetch.

mprynce commented on src/client/auth-handler.ts:
```diff
@@ -0,0 +1,31 @@
+
+    /* If the last call using the headers was successful, report back using this function. */
+    onSuccess: (headers:HttpHeaders) => Promise<void>
+}
```
Fixed :)

mprynce commented:
I've made all the requested changes, and given reasons (I believe) why some of the tests were appropriate.  I've also updated the auth-handler.ts to explain why the onSuccess() call is necessary and the suggeted headers should not be saved until onSuccess is called.

Thank you so much for all the feedback and hoping this branch is ready to merge :)

swapydapy commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,963 @@
+        fetchImpl: mockFetch
+      });
+
+      const messageParams: MessageSendParams = {
```
can we create these as constants and re-use across tests. Same for response for for auth fake. That would substantially improve the readability.

mprynce commented on src/client/client.ts:
```diff
@@ -31,45 +31,53 @@ import {
 type A2AStreamEventData = Message | Task | TaskStatusUpdateEvent | TaskArtifactUpdateEvent;
 
+export interface A2AClientOptions {
+  authHandler?: AuthenticationHandler;
```
Followed the spirit of the above advice and removed the AuthenticationHandler from the A2AClient options and instead wrapped it in a fetchImpl.  Much simpler API now and the fetchImpl can control the retries with dynamic authentication.

mprynce commented:
Did a big refactor, including moving the auth handler out of A2AClient and instead using a fetch wrapper.  Divided client testing into basic testing, and auth testing files.

adamfweidman commented on src/client/index.ts:
```diff
@@ -3,3 +3,4 @@
  */
 
 export { A2AClient } from "./client.js";
```
export A2AClientOptions as well

mprynce replied:
Added export - thank you for the catch!

swapydapy commented on src/client/auth-handler.ts:
```diff
@@ -0,0 +1,125 @@
+
+    let response = await this.fetchImpl(url, mergedInit);
+
+    // Check for HTTP 401/403 and retry request if necessary
```
Rephrase this comment as this AuthHandlingFetch is not checking for 401/403. The authHandler can possibly check for auth errors.

swapydapy commented on src/client/client.ts:
```diff
@@ -35,41 +35,45 @@ import {
 
+export interface A2AClientOptions {
+  agentCardPath?: string;
+  fetchImpl?: typeof fetch;
```
Can we put the fetchImpl in another interface called RequestConfig. This way other options can be added at requestConfig (like timeout). In future, that requestConfig can be overridden per API call as well.

mprynce replied:
I'm looking at client.ts:_fetchRpc(...,requestConfig) which would likely be the recipient of something like a sendMessage(...,requestConfig) and wondering exactly how they would integrate together.

Target is:

    return this.fetchImpl(url, requestInit)

To:

   const fetcher = requestConfig?.fetchImpl ?? this.fetchImpl;
   return fetcher(url, requestInit);


Since fetch doesn't support timeouts, the typical solution is to implement a wrapper such as:

  const fetchWithTimeout = (url, options = {}, timeout = 5000) => {
    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), timeout);

    return fetch(url, {
      ...options,
      signal: controller.signal
    })
    .finally(() => clearTimeout(id)); // Clear the timeout if the fetch completes
};

And this circles us back to more simply creating a fetch wrapper.

Are we getting into scope creep?

swapydapy commented on src/client/auth-handler.ts:
```diff
@@ -0,0 +1,125 @@
+    
+    // Make the instance callable
+    const boundFetch = this._executeFetch.bind(this);
+    Object.setPrototypeOf(boundFetch, AuthHandlingFetch.prototype);
```
This class structure is difficult to comprehend. As suggested in my previous comment, could you refactor this into higher-order-function. That would be much easier to go through and serve as sample for providing custom fetch.

const authenticatedFetch = (fetchImpl: typeof fetch, authHandler: AuthenticationHandler) => {

... return () => {}
}

swapydapy commented on test/client/util.ts:
```diff
@@ -0,0 +1,221 @@
+  const body = JSON.stringify(jsonRpcResponse);
+  
+  // Create a ReadableStream to ensure the body can be read multiple times
+  const stream = new ReadableStream({
```
Any reason to create these streams for generating hardcoded responses. This seems to be generated by LLM.

mprynce replied:
I believe this code was added in response to errors from earlier code.  The defensive design is best practice, but the current tests do not cause this code to be needed, so I have removed it.

swapydapy commented on test/client/client.spec.ts:
```diff
@@ -0,0 +1,360 @@
+}
+
+// Helper function to create fresh mock fetch responses
+function createFreshMockFetch(url: string, options?: RequestInit) {
```
What is the point of this method (naming seems to be incorrect as well). This code can be directly added to `createMockFetch()`

mprynce replied:
This design avoids the issues from response body reuse (the other defensive code removed from a prior change request)

swapydapy commented on test/client/client.spec.ts:
```diff
@@ -0,0 +1,360 @@
+      await client.getAgentCard();
+      
+      // Reset fetch mock
+      mockFetch.reset();
```
No need to reset this. Just check the call count to be 2

mprynce replied:
Removed reset and fixed call count to 1

swapydapy commented on test/client/client.spec.ts:
```diff
@@ -0,0 +1,360 @@
+    });
+  });
+
+  describe('Request ID Management', () => {
```
Not a valid test. Any reason for adding this test.

swapydapy commented on test/server/default_request_handler.spec.ts:
```diff
@@ -208,7 +208,9 @@ describe('DefaultRequestHandler as A2ARequestHandler', () => {
     it('sendMessage: should handle agent execution failure for blocking calls', async () => {
         const errorMessage = 'Agent failed!';
-        (mockAgentExecutor as MockAgentExecutor).execute.rejects(new Error(errorMessage));
+        (mockAgentExecutor as MockAgentExecutor).execute.callsFake(async () => {
```
Not needed

swapydapy reviewed (changes requested):
Thanks @mprynce for the changes. Added comments regarding the structure of AuthHadnlingFetch & tests.

I understand we are using LLMs for assisted coding, but please go through the generated code with fine-toothed comb.

Otherwise the responsibility of making sure there are no bugs lies entirely on the reviewers, which significantly makes adds to their burden.

swapydapy commented on src/client/auth-handler.ts:
```diff
@@ -0,0 +1,125 @@
+ * - const response = await authFetch(url, options);
+ * - const response = await authFetch(url); // Direct function call
+ */
+export class AuthHandlingFetch extends Function {
```
nit: rename to AuthenticatedFetchWithRetry

mprynce replied:
Renamed to createAuthenticatingFetchWithRetry - hope that's ok?

mprynce commented on src/client/auth-handler.ts:
```diff
@@ -0,0 +1,125 @@
+     * Provides additional HTTP request headers.
+     * @returns HTTP headers which should include Authorization if available.
+     */
+    headers: () => HttpHeaders;
```
Good catch on the headers() promise - I've updated that code.  Also updated to onSuccessfulRetry().

Expired credentials is only one authentication failure mode.  Other modes are:
- Client signing a challenge with invalid private keys (outside it's knowledge)
- Client using a signing algorithm that is not supported by the server.
 
For these cases the client **should not assume success**, and instead wait for verification from the server that the Authorization header was accepted.

NOTE that above are outside the centralized authentication server scenarios, and instead use decentralized signing systems.

The onSuccessfulRetry() provides a separation of concerns where the AuthenticationHandler can simply handle (1) cached authorization headers, (2) generating new headers, and (3) caching the headers that worked.  We *could* move that logic into something similar to the AuthHandlingFetch() class, but that overloads that implementation and makes it unnecessarily complex.

mprynce commented on src/client/auth-handler.ts:
```diff
@@ -0,0 +1,125 @@
+    headers: () => HttpHeaders;
+
+    /**
+     * Called to check if the HTTP request should be retried with *new* headers.  This usually
```
Updated to "This method will be always called after each request is executed.  Handler can check if there are auth related failures and if the request needs to be retried with revised headers."

mprynce commented on src/client/auth-handler.ts:
```diff
@@ -0,0 +1,125 @@
+     * function returns new HTTP headers, then the request should be retried with
+     * the revised headers.
+     *
+     * Note that the new headers returned by this request are transient, and will only be saved
```
New wording "Note that the new headers returned by this request may be transient, and might only be saved
     when the onSuccessfulRetry() function is called, or otherwise discarded.  This is an
     implementation detail of an AuthenticationHandler.  If the headers are transient, then
     the onSuccessfulRetry() function should be implemented to save the headers for subsequent
     requests."

mprynce commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,977 @@
+
+
+// Challenge manager class for authentication testing
+class ChallengeManager {
```
This is a real world test and reduces the chances of people adding breaking changes.  I can remove it if you'd like, but we will have greatly reduced coverage.

mprynce replied:
Please confirm you'd like me to remove the ChallengeManager and the related coverage tests - thank you!

mprynce commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,977 @@
+          
+          // If no auth header, return 401 to trigger auth flow
+          if (!authHeader) {
+            const requestId = extractRequestId(options);
```
Soak tests are ok, but I've removed it.

swapydapy commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,762 @@
+    it('should parse WWW-Authenticate header and generate correct Authorization header', async () => {
+      // Create a mock that tracks the Authorization headers sent
+      let capturedAuthHeaders: string[] = [];
+      const authHeaderTestFetch = sinon.stub().callsFake(async (url: string, options?: RequestInit) => {
```
Instead of re-supplying another copy of fetch, with the only difference of capturing auth-headers. Could you do following:
1. Either capture headers in createMockFetch(). All other tests can also use it, and reduces code duplication.
2. Just check args to the mock Fetch stub: stub.getCall(n).args

mprynce replied:
Big refactor of the mock fetch factories into a common createMockFetch that is configurable.  Also attached capturedAuthHeaders to the mock to reduce code duplication

swapydapy commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,762 @@
+    it('should continue without authentication when server does not return 401', async () => {
+      // Create a mock that doesn't require authentication
+      let capturedAuthHeaders: string[] = [];
+      const noAuthRequiredFetch = sinon.stub().callsFake(async (url: string, options?: RequestInit) => {
```
Instead of let's re-use createMockFetch, and an optional param to check for auth. The value here could be False.

mprynce replied:
Yes, made it configurable

swapydapy commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,762 @@
+      }
+    });
+
+    it('should fail gracefully when no authHandler is provided and server returns 401', async () => {
```
We are not failing gracefully here, right. Just conveying the server error. The test should be called "Client pipes server errors when no auth handler is specified"

mprynce replied:
Renamed

swapydapy commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,762 @@
+
+    it('should fail gracefully when no authHandler is provided and server returns 401', async () => {
+      // Create a mock that returns 401 without authHandler
+      const noAuthHandlerFetch = sinon.stub().callsFake(async (url: string, options?: RequestInit) => {
```
nit: fetchWithApiError

mprynce replied:
Renamed

swapydapy commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,762 @@
+          });
+          
+          // Create a Response that can be read multiple times
+          const stream = new ReadableStream({
```
not needed

swapydapy commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,762 @@
+
+  describe('Header Merging', () => {
+    it('should merge auth headers with provided headers', async () => {
+      const authHandlerSpy = sinon.spy(authHandler, 'headers');
```
Don't think auth-handler added any headers, then how is this test asserting the merge behaviour.

swapydapy commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,762 @@
+    });
+  });
+
+  describe('Authentication Retry Logic', () => {
```
This whole is repeat of tests written for A2A client. Please remove these or refactor those test to be at this level, just test that A2AClient uses custom fetch.

swapydapy commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,762 @@
+    });
+  });
+
+  describe('Integration with A2AClient', () => {
```
remove this one

mprynce replied:
Removed

swapydapy commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,762 @@
+      
+      // Mock fetch to return 401 first, then 200
+      let callCount = 0;
+      const successMockFetch = sinon.stub().callsFake(async (url: string, options?: RequestInit) => {
```
Should use existing mockFetch.

swapydapy commented:
The main code looks fine now., though the tests need to cleaned-up.

As mentioned, the tests needs to assert intended library behaviour. The library is general in nature and allows retrying a request when AuthHandler deems a response to be retryable.

It has not specific to one way of auth scheme or failures.

The PR can be merged if you can cleanup the tests. A lot of folks are asking for this auth support. Thank you for your contribution.

mprynce commented:
> The main code looks fine now., though the tests need to cleaned-up.
> 
> As mentioned, the tests needs to assert intended library behaviour. The library is general in nature and allows retrying a request when AuthHandler deems a response to be retryable.
> 
> It has not specific to one way of auth scheme or failures.
> 
> The PR can be merged if you can cleanup the tests. A lot of folks are asking for this auth support. Thank you for your contribution.

Thank you so much for all the attention to detail and great feedback.  I'll get to work on this ASAP!

swapydapy commented:
Let me know the ETA you have in mind for addressing these comments. Might need to send a PR to allow specifying custom fetch in client.

mprynce commented:
> Let me know the ETA you have in mind for addressing these comments. Might need to send a PR to allow specifying custom fetch in client.

Shooting for tonight :)

mprynce commented on src/client/auth-handler.ts:
```diff
@@ -0,0 +1,125 @@
+/**
+ * Generic interface for handling authentication for HTTP requests.
+ * 
+ * An example flow using Universal Authentication ([DID](https://w3c.github.io/did/) based signing
```
Removed Agentic JWT example and replaced with general description

mprynce commented on src/client/auth-handler.ts:
```diff
@@ -0,0 +1,125 @@
+     * New headers are usually needed when the HTTP response issues a 401 or 403.  If this function returns
+     * new HTTP headers, then the request should be retried with the revised headers.
+     *
+     * Note that the new headers returned by this request may be transient, and might only be saved
```
Removed, and already in place in the onSuccessfulRetry() docs

mprynce commented on test/client/util.ts:
```diff
@@ -0,0 +1,210 @@
+  const body = JSON.stringify(data);
+  
+  // Create a ReadableStream to ensure the body can be read multiple times
+  const stream = new ReadableStream({
```
Removed unecessry second createFreshMockFetch, and removed use of ReadableStream when simple string for response body works

mprynce commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,977 @@
+      }
+    });
+
+    it('should parse WWW-Authenticate header and generate correct Authorization header', async () => {
```
Renamed the test to the suggestion and relaxed the test to accept any auth header in second response

mprynce commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,762 @@
+  }
+
+  // We have an auth header, so make sure the scheme is correct
+  const [ scheme, params ] = authHeader.split(/\s+/);
```
Removed Authorization header scheme and token verification from tests

mprynce commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,762 @@
+    const token = ChallengeManager.signChallenge(challenge);
+      
+    // have the client try the token, BUT don't save it in case the client doesn't accept it
+    return { 'Authorization': `Agentic ${token}` };
```
Revised to conventional Bearer

mprynce commented on test/client/client_auth.spec.ts:
```diff
@@ -0,0 +1,762 @@
+      await client.sendMessage(messageParams);
+      
+      // Reset calls
+      mockFetch.reset();
```
Removed reset() and revised callCount to match new values
